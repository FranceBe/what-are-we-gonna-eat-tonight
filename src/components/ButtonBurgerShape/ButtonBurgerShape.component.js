import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { spaces, colors, fontSizes } from 'styles/variables'

export const BurgerShape = styled.button`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-end;
  height: ${spaces.large};
  width: ${spaces.large};
  border-radius: ${spaces.large} ${spaces.large} ${spaces.medium} ${spaces.medium};
  background:
  linear-gradient(0deg,
  ${colors.primary_default} 15%,
  ${colors.secondary_default} 27%,
  ${colors.tertiary_default} 47%,
  ${colors.secondary_light} 45%,
  ${colors.secondary_default} 60%,
  ${colors.primary_default} 79%);
  cursor: pointer;
  :focus {
    outline: none;
  }
`
export const ButtonText = styled.span`
  background-color: rgba(255, 255, 255, 0.7);
  padding: ${spaces.small};
  margin-bottom: ${spaces.small};
  font-size: ${fontSizes.px_14}
`

export class ButtonBurgerShape extends React.PureComponent {
  static propTypes = {
    children: PropTypes.any.isRequired,
    onClick: PropTypes.func,
  }

  onActionClick = e => {
    if (this.props.onClick) {
      this.props.onClick()
    } else {
      e.preventDefault()
    }
  }

  render() {
    const { ...rest } = this.props
    return (
      <BurgerShape onClick={this.onActionClick} {...rest}>
        <ButtonText>{this.props.children}</ButtonText>
      </BurgerShape>
    )
  }
}

export default ButtonBurgerShape
