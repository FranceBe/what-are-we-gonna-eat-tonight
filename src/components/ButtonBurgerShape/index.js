import ButtonBurgerShapeComponent,
{ ButtonBurgerShape, BurgerShape, ButtonText }
  from 'components/ButtonBurgerShape/ButtonBurgerShape.component'

export { ButtonBurgerShape, BurgerShape, ButtonText }

export default ButtonBurgerShapeComponent
