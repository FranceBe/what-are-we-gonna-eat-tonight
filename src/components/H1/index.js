import H1Component, { H1, H1Default } from 'components/H1/H1.component'

export { H1, H1Default }

export default H1Component
