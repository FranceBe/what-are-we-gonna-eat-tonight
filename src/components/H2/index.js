import H2Component, { H2, H2Default } from 'components/H2/H2.component'

export { H2, H2Default }

export default H2Component
