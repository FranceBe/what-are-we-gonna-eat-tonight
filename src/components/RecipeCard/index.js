import RecipeCardComponent, { RecipeCard } from 'components/RecipeCard/RecipeCard.component'

export { RecipeCard }

export default RecipeCardComponent
