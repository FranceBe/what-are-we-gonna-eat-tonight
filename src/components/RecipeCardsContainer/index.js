export { RecipeCardsContainer } from 'components/RecipeCardsContainer/RecipeCardsContainer.component'
