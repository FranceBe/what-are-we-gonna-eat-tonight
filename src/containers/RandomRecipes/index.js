import RandomRecipesContainer,
{ RandomRecipes, ButtonsContainer }
  from 'containers/RandomRecipes/RandomRecipes.container'

export { RandomRecipes, ButtonsContainer }

export default RandomRecipesContainer
