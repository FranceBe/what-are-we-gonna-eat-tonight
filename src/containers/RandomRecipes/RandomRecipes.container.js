import React from 'react'
import styled from 'styled-components'
import H2 from 'components/H2'
import ButtonBurgerShape from 'components/ButtonBurgerShape'
import RecipeCard from 'components/RecipeCard'
import { getRandomInt, getRandomNItems } from 'services/random.service'
import { RecipeCardsContainer } from 'components/RecipeCardsContainer'

export const ButtonsContainer = styled.div`
  display: flex;
  flex-direction: row;
  & button:not(:last-of-type) {
    margin-right: 10px;
  }
`

export class RandomRecipes extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      listOfRecipes: [],
    }
  }

  getRandomRecipes = n => {
    const randomPage = getRandomInt(50)
    fetch(`api/puppy/?p=${randomPage}`)
      .then(res => res.json())
      .then(data => {
        const listOfRecipes = getRandomNItems(data.results, n)
        this.setState({ listOfRecipes })
      })
      .catch(err => {
        throw Error(err)
      })
  }

  render() {
    return (
      <>
        <H2> Combien voulez vous de propositions ? </H2>
        <ButtonsContainer>
          <ButtonBurgerShape onClick={() => this.getRandomRecipes(1)}> 1 </ButtonBurgerShape>
          <ButtonBurgerShape onClick={() => this.getRandomRecipes(5)}> 5 </ButtonBurgerShape>
          <ButtonBurgerShape onClick={() => this.getRandomRecipes(10)}> 10 </ButtonBurgerShape>
        </ButtonsContainer>
        {this.state.listOfRecipes.length > 0 &&
        <RecipeCardsContainer>
          {this.state.listOfRecipes.map((recipe, i) => {
            return (
              <RecipeCard
                key={`recipe-${i}`}
                title={recipe.title}
                ingredients={recipe.ingredients}
                href={recipe.href}
                thumbnail={recipe.thumbnail}
              />
            )
          })}
        </RecipeCardsContainer>
        }
      </>
    )
  }
}

export default RandomRecipes
