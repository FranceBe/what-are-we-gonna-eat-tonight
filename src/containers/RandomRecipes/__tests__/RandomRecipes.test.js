import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import RandomRecipes, { ButtonsContainer } from 'containers/RandomRecipes/RandomRecipes.container'
import { RecipeCardsContainer } from 'components/RecipeCardsContainer'
import H2 from 'components/H2'
import ButtonBurgerShape from 'components/ButtonBurgerShape'
import * as randomService from 'services/random.service'

import RecipeCard from 'components/RecipeCard'

describe('RandomRecipes', () => {
  describe('styled-components', () => {
    describe('ButtonsContainer', () => {
      it('should match snapshot', () => {
        const ButtonsContainerTree = renderer.create(<ButtonsContainer/>).toJSON()
        expect(ButtonsContainerTree).toMatchSnapshot()
      })
    })
  })
  describe('component', () => {
    const listOfRecipes = [
      {
        title: 'Chocolate Cake',
        href: 'http://siteweb.site.web',
        ingredients: 'chocolat, cake, chocolate cake',
        thumbnail: 'https://upload.wikimedia.org/wikipedia/commons/d/df/Chocolate_cake.jpg',
      },
      {
        title: 'Cheese Cake',
        href: 'http://cheese.cake',
        ingredients: 'cheese, cake',
        thumbnail: '',
      }
    ]
    let getRandomIntSpy
    beforeAll(() => {
      getRandomIntSpy =
        jest.spyOn(randomService, 'getRandomInt').mockImplementation(() => 12)
      fetch.mockResponse(JSON.stringify({ results: [...listOfRecipes] }))
    })
    afterAll(() => {
      jest.resetAllMocks()
    })
    describe('rendering', () => {
      it('should render an H2 and a ButtonBurgerShape', () => {
        const RandomRecipesWrapper = shallow(<RandomRecipes/>)
        expect(RandomRecipesWrapper.find(H2)).toHaveLength(1)
        expect(RandomRecipesWrapper.find(H2).dive().text()).toBe(' Combien voulez vous de propositions ? ')
        expect(RandomRecipesWrapper.find(ButtonBurgerShape)).toHaveLength(3)
        expect(RandomRecipesWrapper.find(ButtonBurgerShape).at(0).dive().text())
          .toBe(' 1 ')
        expect(RandomRecipesWrapper.find(ButtonBurgerShape).at(1).dive().text())
          .toBe(' 5 ')
        expect(RandomRecipesWrapper.find(ButtonBurgerShape).at(2).dive().text())
          .toBe(' 10 ')
      })
      it('should call getRandomRecipes when a BurgerButton is clicked', () => {
        const RandomRecipesWrapper = shallow(<RandomRecipes/>)
        const getRandomRecipesSpy = jest.spyOn(RandomRecipesWrapper.instance(), 'getRandomRecipes')
        RandomRecipesWrapper.find(ButtonBurgerShape).at(0).props().onClick()
        expect(getRandomRecipesSpy).toHaveBeenCalledTimes(1)
      })
      it('should render as many RecipeCard as there is objects in listOfRecipes state', () => {
        const RandomRecipesWrapper = shallow(<RandomRecipes/>)
        RandomRecipesWrapper.setState({
          listOfRecipes,
        })
        expect(RandomRecipesWrapper.find(RecipeCardsContainer)).toHaveLength(1)
        const RecipeCardWrapper = RandomRecipesWrapper.find(RecipeCard)
        expect(RecipeCardWrapper).toHaveLength(2)
        expect(RecipeCardWrapper.at(0).props().title).toEqual(listOfRecipes[0].title)
        expect(RecipeCardWrapper.at(0).props().ingredients).toEqual(listOfRecipes[0].ingredients)
        expect(RecipeCardWrapper.at(0).props().href).toEqual(listOfRecipes[0].href)
        expect(RecipeCardWrapper.at(0).props().thumbnail).toEqual(listOfRecipes[0].thumbnail)
        expect(RecipeCardWrapper.at(1).props().title).toEqual(listOfRecipes[1].title)
        expect(RecipeCardWrapper.at(1).props().ingredients).toEqual(listOfRecipes[1].ingredients)
        expect(RecipeCardWrapper.at(1).props().href).toEqual(listOfRecipes[1].href)
        expect(RecipeCardWrapper.at(1).props().thumbnail).toEqual(listOfRecipes[1].thumbnail)
      })
    })
    describe('methods', () => {
      describe('getRandomRecipes', () => {
        it('should call getRandomInt', () => {
          const RandomRecipesWrapper = shallow(<RandomRecipes/>)
          RandomRecipesWrapper.instance().getRandomRecipes()
          expect(getRandomIntSpy).toHaveBeenCalledTimes(1)
          expect(getRandomIntSpy).toHaveReturnedWith(12) // output is mocked in beforeAll
        })
        it('should call fetch', () => {
          const RandomRecipesWrapper = shallow(<RandomRecipes/>)
          RandomRecipesWrapper.instance().getRandomRecipes()
          expect(fetch).toHaveBeenCalledTimes(1)
          expect(fetch).toHaveBeenCalledWith("api/puppy/?p=12")
        })
      })
    })
  })
})
