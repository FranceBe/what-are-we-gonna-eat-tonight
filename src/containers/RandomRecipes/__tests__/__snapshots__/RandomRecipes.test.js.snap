// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomRecipes styled-components ButtonsContainer should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 button:not(:last-of-type) {
  margin-right: 10px;
}

<div
  className="c0"
/>
`;
